flow/connectors/mongo/mongo.go

```diff
@@ -0,0 +1,433 @@
+	})
+	if err != nil {
+		return model.SetupReplicationResult{}, fmt.Errorf("failed to store initial resume token: %w", err)
+	}
+	c.logger.Info("SetupReplication completed, stored initial resume token",
+		slog.String("flowJobName", input.FlowJobName))
+	return model.SetupReplicationResult{}, nil
+}
+
+// stubs for CDCPullConnectorCore
+
+func (c *MongoConnector) EnsurePullability(ctx context.Context, req *protos.EnsurePullabilityBatchInput) (
+	*protos.EnsurePullabilityBatchOutput, error,
+) {
+	return nil, nil
+}
+
+func (c *MongoConnector) ExportTxSnapshot(context.Context, map[string]string) (*protos.ExportTxSnapshotOutput, any, error) {
+	return nil, nil, nil
+}
+
+func (c *MongoConnector) FinishExport(any) error {
+	return nil
+}
+
+func (c *MongoConnector) SetupReplConn(context.Context) error {
+	return nil
+}
+
+func (c *MongoConnector) ReplPing(context.Context) error {
+	return nil
+}
+
+func (c *MongoConnector) UpdateReplStateLastOffset(ctx context.Context, lastOffset model.CdcCheckpoint) error {
+	return nil
+}
+
+func (c *MongoConnector) PullFlowCleanup(ctx context.Context, jobName string) error {
+	return nil
+}
+
+func (c *MongoConnector) HandleSlotInfo(
+	ctx context.Context,
+	alerter *alerting.Alerter,
+	catalogPool shared.CatalogPool,
+	alertKeys *alerting.AlertKeys,
+	slotMetricGauges otel_metrics.SlotMetricGauges,
+) error {
+	return nil
+}
+
+func (c *MongoConnector) GetSlotInfo(ctx context.Context, slotName string) ([]*protos.SlotInfo, error) {
+	return nil, nil
+}
+
+func (c *MongoConnector) AddTablesToPublication(ctx context.Context, req *protos.AddTablesToPublicationInput) error {
+	return nil
+}
+
+func (c *MongoConnector) RemoveTablesFromPublication(ctx context.Context, req *protos.RemoveTablesFromPublicationInput) error {
+	return nil
+}
+
+// end stubs
+
+func (c *MongoConnector) PullRecords(
+	ctx context.Context,
+	catalogPool shared.CatalogPool,
+	otelManager *otel_metrics.OtelManager,
+	req *model.PullRecordsRequest[model.RecordItems],
+) error {
+	defer req.RecordStream.Close()
+	c.logger.Info("[started] PullRecords for mirror "+req.FlowJobName,
+		slog.Any("table_mapping", req.TableNameMapping),
+		slog.Uint64("max_batch_size", uint64(req.MaxBatchSize)),
+		slog.Duration("idle_timeout", req.IdleTimeout))
+
+	changeStreamOpts := options.ChangeStream().
+		SetComment("PeerDB changeStream for mirror " + req.FlowJobName).
+		SetFullDocument(options.UpdateLookup).
+		SetFullDocumentBeforeChange(options.WhenAvailable)
+	if req.LastOffset.Text != "" {
+		// If we have a last offset, we resume from that point
+		c.logger.Info("[mongo] resuming change stream", slog.String("resumeToken", req.LastOffset.Text))
+		resumeTokenBytes, err := base64.StdEncoding.DecodeString(req.LastOffset.Text)
+		if err != nil {
+			return fmt.Errorf("failed to parse last offset: %w", err)
+		}
+		changeStreamOpts.SetResumeAfter(bson.Raw(resumeTokenBytes))
+	}
+
+	changeStream, err := c.client.Watch(ctx, mongo.Pipeline{}, changeStreamOpts)
```

Comment:
Reading this code, it looks like we are creating a new changestream every time `PullRecords` gets called. I think ideally we'd want to create a long-lived changestream object once, and call changestream.Next(...) on it each time `PullRecords` is called.  There's concerns around having too many changestream opened against the database.

Reply:
ah, according to the link:
> Each change stream uses a connection and a [getMore](https://www.mongodb.com/docs/manual/reference/command/getMore/#std-label-manual-reference-commands-getMore) operation on the change stream for the period of time that it waits for the next event.

I think this may be the source of `resumeToken` when there's no change event.

Reply:
tracked: https://github.com/PeerDB-io/peerdb/issues/3106